framework/src/utils/MooseCoordTransform.C

```diff
@@ -0,0 +1,338 @@
+      "rz_coord_axis", rz_coord_axis, "The rotation axis (X | Y) for axisymetric coordinates");
+  params.addParam<Real>("length_units_per_meter",
+                        "How many mesh length units are in a meter, e.g. if your mesh units are "
+                        "centimeters, then this parameter value should be 100.");
+  params.addRangeCheckedParam<Real>(
+      "alpha_rotation",
+      "-180<alpha_rotation<=180",
+      "The number of degrees that the domain should be alpha-rotated using the Euler "
+      "angle ZXZ convention from https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix in "
+      "order to align with a canonical physical space of your choosing.");
+  params.addRangeCheckedParam<Real>(
+      "beta_rotation",
+      "-180<beta_rotation<=180",
+      "The number of degrees that the domain should be beta-rotated using the Euler "
+      "angle ZXZ convention from https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix in "
+      "order to align with a canonical physical space of your choosing.");
+  params.addRangeCheckedParam<Real>(
+      "gamma_rotation",
+      "-180<gamma_rotation<=180",
+      "The number of degrees that the domain should be gamma-rotated using the Euler "
+      "angle ZXZ convention from https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix in "
+      "order to align with a canonical physical space of your choosing.");
+  MooseEnum up_direction("X=0 Y=1 Z=2");
+  params.addParam<MooseEnum>(
+      "up_direction",
+      up_direction,
+      "Specify what axis corresponds to the up direction in physical space (the opposite of the "
+      "gravity vector if you will). If this parameter is provided, we will perform a single 90 "
+      "degree rotation of the domain--if the provided axis is 'x' or 'z', we will not rotate if "
+      "the axis is 'y'--such that a point which was on the provided axis will now lie on the "
+      "y-axis, e.g. the y-axis is our canonical up direction. If you want finer grained control "
+      "than this, please use the 'alpha_rotation', 'beta_rotation', and 'gamma_rotation' "
+      "parameters.");
+  return params;
+}
+
+MooseCoordTransform::MooseCoordTransform(const InputParameters & params)
+{
+  //
+  // coordinate system transformation. If we have multiple different coordinate system types in our
+  // problem, we take note of it because that can cause issues if there is a non-Cartesian
+  // destination coordinate system
+  //
+  const auto & mesh = *params.getCheckedPointerParam<MooseMesh *>("mesh");
+  const auto & coord_sys = mesh.getCoordSystem();
+  std::unordered_set<Moose::CoordinateSystemType> coord_types;
+  auto map_it = coord_sys.begin();
+  // It's possible that the mesh is not in a complete state
+  if (map_it == coord_sys.end())
+    setCoordinateSystem(Moose::COORD_XYZ);
+  else
+    setCoordinateSystem(
+        map_it->second,
+        Direction(static_cast<unsigned int>(int(params.get<MooseEnum>("rz_coord_axis")))));
```

Comment:
So you always set the _coord_type to be the first coordinate type entry in the `coord_sys` map? But down below in the `setDestinationCoordinateSystem` method, you are checking whether the destination coord transformation `_has_different_coord_sys`.

Reply:
We check *our* `_has_different_coord_sys` but you're right, I should should check `destination_coord_transform._has_different_coord_sys` as well